### PR TITLE
mam: api serialize

### PR DESCRIPTION
**Pull request by tsvisabo**
- Serialize mam_api_t struct

### CODE REVIEW

File: mam/api/api.c [hunk 1]
```diff
@@ -369,3 +382,206 @@ retcode_t mam_api_bundle_read_packet(mam_api_t *const api,
...
+  }
+}
+
+retcode_t mam_api_send_ctx_map_deserialize(
+    trits_t const *const buffer,
+    trit_t_to_mam_msg_send_context_t_map_t *const map) {
+  retcode_t ret;
+  mam_msg_send_context_t ctx;
+  trit_t msg_id[MAM_MSG_ID_SIZE];
+  const size_t num_ctxs = 0;
```
Comment (thibault-martinez) [inline mam/api/api.c]: Why const if we are decoding into ?

File: mam/api/api.c [hunk 2]
```diff
@@ -369,3 +382,206 @@ retcode_t mam_api_bundle_read_packet(mam_api_t *const api,
...
+  return RC_OK;
+}
+
+retcode_t mam_api_recv_ctx_map_deserialize(
+    trits_t const *const buffer,
+    trit_t_to_mam_msg_recv_context_t_map_t *const map) {
+  retcode_t ret;
+  mam_msg_recv_context_t ctx;
+  trit_t msg_id[MAM_MSG_ID_SIZE];
+  const size_t num_ctxs = 0;
```
Comment (thibault-martinez) [inline mam/api/api.c]: Why const if we are decoding into ?

File: mam/api/api.c [hunk 3]
```diff
@@ -369,3 +382,206 @@ retcode_t mam_api_bundle_read_packet(mam_api_t *const api,
...
+    HASH_ITER(hh, api->channels, curr_channel_entry, tmp_channel_entry) {
+      if (memcmp(curr_channel_entry->value.mss.root,
+                 curr_ctx_entry->value.mss_root, MAM_CHANNEL_ID_SIZE) == 0) {
+        curr_ctx_entry->value.mss = &curr_channel_entry->value.mss;
+      } else {
+        HASH_ITER(hh, curr_channel_entry->value.endpoints, endpoint_entry,
+                  tmp_endpoint_entry) {
+          if (memcmp(endpoint_entry->value.mss.root,
+                     curr_ctx_entry->value.mss_root,
+                     MAM_CHANNEL_ID_SIZE) == 0) {
```
Comment (thibault-martinez) [inline mam/api/api.c]: Same size but `MAM_ENDPOINT_ID_SIZE `

File: mam/mam/message.h [hunk 4]
```diff
@@ -54,6 +54,7 @@ typedef struct mam_msg_send_context_s {
   mam_spongos_t spongos;
   trint18_t ord;
   mam_mss_t *mss;
+  trit_t mss_root[MAM_MSS_PK_SIZE];
```
Comment (thibault-martinez) [inline mam/mam/message.h:57]: In https://github.com/iotaledger/entangled/pull/911 the root is now serialized with the MSS so no need for this field

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/api/api.c (hunk 1 above)
Comment (tsvisabo) [reply mam/api/api.c]: right

File: mam/api/api.c (hunk 2 above)
Comment (tsvisabo) [reply mam/api/api.c]: right

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**